pkg/chunk/chunk_store.go

```diff
@@ -411,38 +433,48 @@ func (c *Store) lookupChunksByMetricName(ctx context.Context, from, through mode
 				incomingErrors <- err
 				return
 			}
+			level.Debug(logger).Log("func", "ChunkStore.lookupChunksByMetricName", "matcher", matcher, "queries", len(queries))
 
 			// Lookup IndexEntry's
 			entries, err := c.lookupEntriesByQueries(ctx, queries)
 			if err != nil {
 				incomingErrors <- err
 				return
 			}
+			level.Debug(logger).Log("func", "ChunkStore.lookupChunksByMetricName", "matcher", matcher, "entries", len(entries))
 
-			// Convert IndexEntry's into chunks
-			chunks, err := c.convertIndexEntriesToChunks(ctx, entries, matcher)
+			// Convert IndexEntry's to chunk IDs, filter out non-matchers at the same time.
+			chunkIDs, err := c.parseIndexEntries(ctx, entries, matcher)
 			if err != nil {
 				incomingErrors <- err
-			} else {
-				incomingChunkSets <- chunks
+				return
 			}
+			level.Debug(logger).Log("func", "ChunkStore.lookupChunksByMetricName", "matcher", matcher, "chunkIDs", len(chunkIDs))
+			incomingChunkIDs <- chunkIDs
 		}(matcher)
 	}
 
 	// Receive chunkSets from all matchers
-	var chunkSets []ByKey
+	var chunkIDSets [][]string
 	var lastErr error
 	for i := 0; i < len(matchers); i++ {
 		select {
-		case incoming := <-incomingChunkSets:
-			chunkSets = append(chunkSets, incoming)
+		case incoming := <-incomingChunkIDs:
+			chunkIDSets = append(chunkIDSets, incoming)
 		case err := <-incomingErrors:
 			lastErr = err
 		}
 	}
+	if lastErr != nil {
+		return nil, lastErr
+	}
+
+	// Merge entries in order because we wish to keep label series together consecutively
+	chunkIDs := nWayIntersectStrings(chunkIDSets)
```

Comment:
Rather than assembling a list then breaking it down to intersect, I'd think it is cheaper to do the intersecting as the results come in. But that could be a subsequent change.